lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,412 @@
+            msg = f"Error creating k8s client: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+            return
+
+        except ApiError as e:
+            if e.status.code == 403:
+                msg = f"Kubernetes resources patch failed: `juju trust` this application. {e}"
+            else:
+                msg = f"Kubernetes resources patch failed: {e}"
+
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+
+        except ValueError as e:
+            msg = f"Kubernetes resources patch failed: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+
+        else:
+            logger.info(
+                "Kubernetes resources for app '%s', container '%s' patched successfully: %s",
+                self._app,
+                self._container_name,
+                self.resource_reqs,
+            )
+
+    @classmethod
+    def _conv_res_req(cls, res_req: ResourceRequirements) -> ResourceRequirements:
+        """Convert ResourceRequirements to comparable form.
+
+        - Convert "memory" to GiB.
+        - Convert "cpu" to float.
+
+        When patching the StatefulSet with {"memory": "0.9Gi"}, the actual PodSpec has
+        {"memory": "966367641600m"}; similarly, {"cpu": "0.30000000000000004"} -> {"cpu": "301m"}.
+        So need to parse the strings and convert before comparing.
+        """
+
+        def _conv(dct: ResourceSpecDict) -> Optional[ResourceSpecDict]:
+            """Convert the memory value of a ResourceSpecDict to GiB representation.
+
+            Raises:
+                ValueError, for invalid input.
+            """
+            if not dct:
+                return None
+            copy = dct.copy()
+
+            if memory := copy.get("memory"):
+                if memory.endswith("m"):
+                    # This is milli. Divide by 1000.
+                    value = bitmath.Byte(float(memory[:-1]) / 1000).to_GiB().value
+                else:
+                    if not memory.endswith("B"):
+                        # Bitmath doesn't recognize e.g. Gi/G - needs to be GiB/GB.
+                        memory += "B"
+                    value = bitmath.parse_string(memory).to_GiB().value
+
+                if value < 0:
+                    raise ValueError(
+                        "Failed to apply memory resource limit patch: "
+                        "value must be greater than or equal to 0"
+                    )
+                copy["memory"] = f"{str(value)}Gi"
+
+            if cpu := copy.get("cpu"):
+                # TODO need to take into account m, k/K, M, G, T, P and E, but for CPU count it's
+                # probably ok to only support "m" and plain decimals for now.
+                # https://github.com/gtsystem/lightkube/issues/36
+                if cpu.endswith("m"):
+                    # This is milli. Divide by 1000.
+                    value = int(cpu[:-1]) / 1000.0
```

Comment:
May want to extend that comment above to explicitly state if loss of precision is intended, as it seems.

Reply:
Yes, K8s spec is round up to millis. But all this will go away when using `parse_quantity`.